### PR TITLE
[Kernel] FP8 support for MoE kernel / Mixtral

### PR DESCRIPTION
**Pull request by pcmoritz**
This PR is the first step towards fixing https://github.com/vllm-project/vllm/pull/3208

It implements dynamic per-tensor scaling (see https://github.com/vllm-project/vllm/pull/4118), so users do not need to compute activation scales on a calibration dataset and they also don't need to convert their model checkpoints. It is enough to specify the `quantization="fp8"` argument. You can try out the PR like this:

```python
from vllm import LLM, SamplingParams

prompts = [
    "Hello, my name is",
    "The president of the United States is",
    "The capital of France is",
    "The future of AI is",
]
sampling_params = SamplingParams(temperature=0.8, top_p=0.95)

llm = LLM(model="mistralai/Mixtral-8x7B-Instruct-v0.1", tensor_parallel_size=2, quantization="fp8")

outputs = llm.generate(prompts, sampling_params)

# Print the outputs.
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
```

**Performance**: For this PR, the focus is on making the code clean (while still trying to get reasonable performance), there is a bunch of optimizations that we will submit as a follow up PR that significantly improve the performance (similar to the numbers in https://github.com/vllm-project/vllm/pull/3954). With this PR, the results are as follows:

Screenshot 2024-04-21 at 1 31 50 PM

**Accuracy**: The accuracy with this PR on MMLU on `mistralai/Mixtral-8x7B-v0.1` is as follows:

```
|      Groups      |Version|Filter|n-shot|Metric|Value |   |Stderr|
|------------------|-------|------|-----:|------|-----:|---|-----:|
|mmlu              |N/A    |none  |     0|acc   |0.7018|±  |0.0036|
| - humanities     |N/A    |none  |     5|acc   |0.6472|±  |0.0065|
| - other          |N/A    |none  |     5|acc   |0.7673|±  |0.0072|
| - social_sciences|N/A    |none  |     5|acc   |0.8099|±  |0.0070|
| - stem           |N/A    |none  |     5|acc   |0.6131|±  |0.0083|
```
this compares favorably with the fp16 results which are
```
|      Groups      |Version|Filter|n-shot|Metric|Value |   |Stderr|
|------------------|-------|------|-----:|------|-----:|---|-----:|
|mmlu              |N/A    |none  |     0|acc   |0.7020|±  |0.1313|
| - humanities     |N/A    |none  |     5|acc   |0.6425|±  |0.1349|
| - other          |N/A    |none  |     5|acc   |0.7744|±  |0.1038|
| - social_sciences|N/A    |none  |     5|acc   |0.8131|±  |0.0695|
| - stem           |N/A    |none  |     5|acc   |0.6108|±  |0.1383|
```

Happy hacking!

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Review by comaniac — COMMENTED** [review]
Overall LGTM!
Leave some comments for interface but open to discuss.

File: vllm/model_executor/models/mixtral.py [hunk 1]
```diff
@@ -66,13 +68,15 @@ def __init__(
         intermediate_size: int,
         params_dtype: Optional[torch.dtype] = None,
         tp_size: Optional[int] = None,
+        use_fp8: bool = True,
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply vllm/model_executor/models/mixtral.py:71]: `mixtral_quant.py` is currently a bit of a hack for the quantized models that don't support the fused_moe kernel. Since with fp8 we support the fused_moe kernel, we can use `mixtral.py` -- hopefully in the future we can get rid of `mixtral_quant.py`. I agree with your suggestion to do something like `quantize_method`, will try that out :)

File: vllm/model_executor/models/mixtral.py [hunk 2]
```diff
@@ -171,6 +200,12 @@ def __init__(self,
         self.rope_theta = rope_theta
         self.sliding_window = sliding_window
 
+        if isinstance(linear_method, Fp8LinearMethod):
+            # If we are using FP8, we currently do not want to
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply vllm/model_executor/models/mixtral.py]: Done!

File: vllm/model_executor/models/mixtral.py [hunk 3]
```diff
@@ -97,6 +101,16 @@ def __init__(
                         device="cuda",
                         dtype=self.params_dtype))
 
+        # Scaling factors for fp8 weights
+        self.ws_scale = nn.Parameter(torch.ones(self.num_total_experts,
+                                                device="cuda",
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply vllm/model_executor/models/mixtral.py:107]: Yeah, for now I just did the same thing that we do for the weights to keep the amount of changes small, let me try to remove this in a follow up PR :)

File: vllm/model_executor/models/mixtral.py [hunk 4]
```diff
@@ -118,6 +133,18 @@ def weight_loader(self, param: nn.Parameter, loaded_weight: torch.Tensor,
...
+        if self.use_fp8:
+            ws = torch.empty_like(self.ws.data, dtype=torch.float8_e4m3fn)
+            w2s = torch.empty_like(self.w2s.data, dtype=torch.float8_e4m3fn)
+            for expert in range(self.num_total_experts):
+                ws[expert, :, :], self.ws_scale[expert] = per_tensor_quantize(
+                    self.ws.data[expert, :, :])
+                w2s[expert, :, :], self.w2s_scale[
+                    expert] = per_tensor_quantize(self.w2s.data[expert, :, :])
+            self.ws = nn.Parameter(ws, requires_grad=False)
+            self.w2s = nn.Parameter(w2s, requires_grad=False)
```
Comment (mgoin) [inline vllm/model_executor/models/mixtral.py:148]: if we define cuda device for the original Parameter, should we do the same here?

File: vllm/model_executor/models/mixtral.py [hunk 5]
```diff
@@ -238,7 +273,8 @@ def __init__(
             num_experts=config.num_local_experts,
             top_k=config.num_experts_per_tok,
             hidden_size=config.hidden_size,
-            intermediate_size=config.intermediate_size)
+            intermediate_size=config.intermediate_size,
+            use_fp8=isinstance(linear_method, Fp8LinearMethod))
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply vllm/model_executor/models/mixtral.py]: Fixed!

File: vllm/model_executor/models/mixtral.py (hunk 4 above)
Comment (pcmoritz) [reply vllm/model_executor/models/mixtral.py:148]: `ws` will inherit the device from `self.ws.data` through `torch.empty_like`, so I don't think we need to specify a CUDA device here (and we do want to remove it from the original Parameter going forward too)

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 6]
```diff
@@ -240,18 +262,21 @@ def invoke_fused_moe_kernel(A: torch.Tensor, B: torch.Tensor, C: torch.Tensor,
...
+        use_fp8=use_fp8,
         **config,
     )
 
 
-def get_config_file_name(E: int, N: int) -> str:
+def get_config_file_name(E: int, N: int, dtype: Optional[str]) -> str:
     device_name = torch.cuda.get_device_name().replace(" ", "_")
-    return f"E={E},N={N},device_name={device_name}.json"
+    dtype_selector = "" if not dtype else f",dtype={dtype}"
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:273]: It turns out, this actually broke the loading since if we do this, Python will add an additional set of quotes (i.e. `dtype='fp8'` instead of `dtype=fp8` which breaks loading the file. Since this is a file name, it is better not to have those quotes I think, so I just reverted this :)

**Review by comaniac — APPROVED** [review]
LGTM!
btw, the scaled quantization kernel shows 2-3x speedup over the PyTorch op implementation in FP8 linear method on my L4, so we should expect some improvements if we use this op in the FP8 linear method. I'll send a follow-up PR after this.

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 7]
```diff
@@ -129,7 +136,10 @@ def fused_moe_kernel(
                     mask=offs_k[:, None] < K - k * BLOCK_SIZE_K,
                     other=0.0)
         # We accumulate along the K dimension.
-        accumulator += tl.dot(a, b)
+        if use_fp8:
+            accumulator = tl.dot(a, b, acc=accumulator, allow_tf32=True)
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:140]: nit: allow_tf is [deprecated](https://github.com/openai/triton/blob/7c6c2ef26430d34bfec657d6e2d957505425e577/python/triton/language/core.py#L1512) and it is recommended to use `input_precision="tf32"`
Also it might be worth noting if this is either: required for fp8 to work or just a performance optimization

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 8]
```diff
@@ -207,22 +220,31 @@ def moe_align_block_size(
...
-                            config: Dict[str, Any]) -> None:
+                            config: Dict[str, Any], compute_type: tl.dtype,
+                            use_fp8: bool) -> None:
     assert topk_weights.stride(1) == 1
     assert sorted_token_ids.stride(0) == 1
 
+    if use_fp8:
+        A, A_scale = ops.scaled_fp8_quant(A)
+    else:
+        A_scale = None
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:237]: Worth asserting that B_scale is None as well
```suggestion
        assert B_scale is None
        A_scale = None
```

File: vllm/model_executor/models/mixtral.py [hunk 9]
```diff
@@ -171,6 +203,12 @@ def __init__(self,
         self.rope_theta = rope_theta
         self.sliding_window = sliding_window
 
+        if isinstance(linear_method, Fp8LinearMethod):
+            # FIXME(pcmoritz): If we are using FP8, we currently do
+            # not want to quantize the attention layers until we improve
+            # the performance and make sure the accuracy is good.
+            linear_method = None
```
Comment (mgoin) [inline vllm/model_executor/models/mixtral.py:210]: Nit: It would be nice to have this as a log since my assumption is all Linear modules would be quantized like in the initial FP8 PR, but I assume this will be a fast follow to fix

**Review by mgoin — COMMENTED** [review]
Nice job, looks good to me!

File: vllm/model_executor/layers/fused_moe/fused_moe.py (hunk 7 above)
Comment (pcmoritz) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:140]: So the `input_precision="tf32"` interface is not part of the triton 2.2.0 release yet (which is the version we are using -- triton is pinned to 2.2.0 via the pytorch version we are pinning).

I tried without this flag and can't measure a performance difference, so will remove it. It seems it is the default for devices with tensor cores anyways (https://triton-lang.org/main/python-api/generated/triton.language.dot.html#triton.language.dot). Removing the flag will make us robust against the `allow_tf32` vs. `input_precision` interface change as well. Once we have upgraded trition, it might be worth looking at https://github.com/openai/triton/pull/3234

File: vllm/model_executor/models/mixtral.py (hunk 9 above)
Comment (pcmoritz) [reply vllm/model_executor/models/mixtral.py]: Fixed :)

File: vllm/model_executor/layers/fused_moe/fused_moe.py (hunk 8 above)
Comment (pcmoritz) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:235]: Thanks, I fixed it :)

**Review by cadedaniel — APPROVED** [review]
Great work! Approving after a light pass given @comaniac and @mgoin's approval.

One question:
> It implements dynamic per-tensor scaling (see https://github.com/vllm-project/vllm/pull/4118), so users do not need to compute activation scales on a calibration dataset and they also don't need to convert their model checkpoints.

Doesn't this approach still depend on the choice of scaling factor as not all datasets will have the same activation patterns for a given layer? Perhaps the d(quality)/d(per-tensor scaling factor) is small versus other methods, but there is still some impact on quality by the chosen per-tensor scaling factor.

**Comment by pcmoritz** [comment]
So in a way it will select the best scaling factor possible at runtime (assuming per-tensor and not subdividing the tensor further for scaling), since it will use the activations in the forward pass to compute the scaling factor. The resulting fp8 tensor will always have 448.0 as its maximum and you can't quantize a given tensor better than that.

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 10]
```diff
@@ -129,7 +136,10 @@ def fused_moe_kernel(
                     mask=offs_k[:, None] < K - k * BLOCK_SIZE_K,
                     other=0.0)
         # We accumulate along the K dimension.
-        accumulator += tl.dot(a, b)
+        if use_fp8:
+            accumulator = tl.dot(a, b, acc=accumulator)
```
Comment (pingzhuu) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:140]: Thanks for your great work! I have a question here, why does fp8 need to do this, is there any difference?

Comment (pcmoritz) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:140]: This is a great question, I'm doing it because it supports the fp8 fast accumulation, see

https://github.com/NVIDIA/cutlass/blob/5c447dd84f8ae0e1d48ff9a2eae26ce8c4958101/CHANGELOG.md?plain=1#L67

https://github.com/openai/triton/blob/5623cdc5fb2d497d2d48cea89170707a97029219/python/triton/ops/matmul.py#L132

The downside is that you can't do rescaling before adding it to the accumulator
